stdlib/public/core/UnicodeHelpers.swift

```diff
@@ -190,11 +190,44 @@ internal func _numUTF8CodeUnits(_ scalar: Unicode.Scalar) -> Int {
     default: return 4
   }
 }
+
 @inline(__always)
 internal func _numUTF16CodeUnits(_ scalar: Unicode.Scalar) -> Int {
   return scalar.value <= UInt16.max ? 1 : 2
 }
 
+@inline(__always) @inlinable
+internal func _mayNotBeNFCStarter(_ byte:UInt8) -> Bool {
+  // Fast-path: All scalars up through U+02FF are NFC and have boundaries
+  // before them. UTF8 encoding of >=U+0300 has a first byte >= 0XCC.
+  return _slowPath(byte >= 0xCC)
+}
+
+@inlinable
+internal func _fastIsSingleByteGrapheme(
+  _ utf8: UnsafeBufferPointer<UInt8>, at i: Int
+  ) -> Bool {
+  let count = utf8.count
+  _internalInvariant(i < count)
+  
+  let byte = utf8[_unchecked: i]
+  guard _fastPath(_isASCII(byte)) else {
+    return false
+  }
+  
+  guard _fastPath(i != count &- 1) else {
+    return true
+  }
+
+  let secondByte = utf8[_unchecked: i &+ 1]
+
+  guard !_mayNotBeNFCStarter(secondByte) else {
```

Comment:
This name is still bad, since "may not" is ambiguous, and both places it's used immediately invert it. Maybe _nonexhaustivelyIsNFCStarter()?